### ISSUE TITLE
Fix S3928 FP: Ambient method parameters in lambdas

### ISSUE THREAD

**Issue by hankovich** [SonarSource/sonar-dotnet#7630]
### Description

S3928 is reported in lambdas for ambient method parameters

### Repro steps

```c#
class Observer
{
    private readonly ConcurrentDictionary<string, int> _dictionary = new();

    public void Observe(string key, int value, Behavior behavior = Behavior.Append)
    {
        _dictionary.AddOrUpdate(key, value, (_, oldValue) => behavior switch
        {
            Behavior.Append => oldValue + value,
            Behavior.Override => value,
            _ => throw new ArgumentOutOfRangeException(nameof(behavior), behavior, null), // <- "behavior" is not declared in the argument list
        });
    }
}

enum Behavior
{
    Append,
    Override,
}
```

### Expected behavior

No diagnostic is provided

### Actual behavior

Diagnostic is reported

### Related information

* C#/VB.NET Plugins version 9.5.0.73987
* Visual Studio version latest
* MSBuild / dotnet version .NET 8 Preview 6

**Comment by sebastien-marichal** [comment]
Hello @hankovich,

Thank you for your feedback.

This is considered a true positive by our analyzer.

A lambda execution is postponed (especially in asynchronous execution, see S4457 for a concrete example), and so are any checks inside.
Since the parameter is captured, the check can be done before the lambda definition and execution, adopting a fail-fast approach.

One compliant implementation to your example could be:
```cs
class Observer
{
    private readonly ConcurrentDictionary<string, int> _dictionary = new();

    public void Observe(string key, int value, Behavior behavior = Behavior.Append)
    {
        if (!Enum.IsDefined(typeof(Behavior), behavior))
        {
            throw new ArgumentOutOfRangeException(nameof(behavior), behavior, null);
        }

        _dictionary.AddOrUpdate(key, value, (_, oldValue) => behavior switch
        {
            Behavior.Append => oldValue + value,
            Behavior.Override => value,
            _ => throw new UnreachableException() // Unreachable since the check is happening before
        });
    }
}
```

Hoping this helps!
